### DISCUSSION TITLE
Can lanceDB effectively handle the scenario of randomly selecting mini batches?

### DISCUSSION THREAD

**Discussion by shanzhongkai** [lancedb/lancedb#2571] [Q&A]
In my understanding, during model training, a mini batch needs to be randomly selected for training in each epoch. If lanceDB is used to implement this, would it become N random queries? Or is there a faster implementation? thanks~

**Comment by shanzhongkai** [comment]
If the samples are stored in HDFS, randomly selecting N rows of samples would require N rounds of RPC, wouldn't that be too slow? What is the correct way to use it?

**Comment by westonpace** [comment]
I'm not entirely sure what you're asking.  I'm currently working on data loading capabilities for LanceDB (e.g. https://github.com/lancedb/lancedb/pull/2552) and there are a number of things that might be referred to as "batches".

> In my understanding, during model training, a mini batch needs to be randomly selected for training in each epoch.

There are several operations that happen during training.

* Test / Training split - The dataset is sometimes divided into a "test split" and a "training split"  For example, we might want 5% of our data to be test data and 95% to be training data.  The split definitions do _not_ typically change from epoch to epoch.  However, some techniques like k-fold cross validation create many different splits, but here you change the split between each model and not between each epoch.

* Iterative training - Training a very large dataset typically requires an iterative training process where you sample the dataset in batches.  This helps control the memory usage of the training process.

* Distributed training - Distributed training distributes a dataset into multiple partition where each worker processes a single partition of the data.

* "Clumped" shuffling - Training typically requires the input is randomized (shuflfed) to avoid over-learning patterns in the data.  However, pure row-by-row shuffling can lead to an excess of random access.  To avoid this some tools allow you to shuffle the data in small "clumps" of rows where each clump stays in order.  For example, if you shuffle in clumps of 16 rows then you will have 16 times fewer IOPS but also less randomization.

It sounds like you are most interested in the random access cost.   In order to randomly query N rows and M columns in LanceDB you need to perform at least N * M IOPS (possibly more depending on the data type & size).  If you want to reduce this load you can _pack_ columns together.  Packing M columns means you can read those M columns in 1 IOP instead of M IOPS.  Clumping will help reduce the IOPS needed as well.  Caching is also very useful if you're training against cloud storage, especially when you're running multiple epochs.

**Reply by shanzhongkai** [reply]
great! thanks for your answer, as you said, I am indeed most interested in random access cost, our scene is 'Clumped' shuffle.

We will first download the samples (line storage) from cloud storage to the training machine, then load them into memory for global random fragmentation and finally read. This approach puts a lot of memory pressure, but the ultimate goal is to extract a mini batch during each training round. 

So, I am thinking about how to achieve this requirement if we use lanceDB? What is the most common way to read samples that need to be stored in cloud storage? thanks~

PS: Our data is all multimodal data
